gnosis_vpn-ctl/src/main.rs

```diff
@@ -1,42 +1,81 @@
-use gnosis_vpn_lib::{command, socket};
+use exitcode::{self, ExitCode};
+use std::process;
 
-mod cli;
+use gnosis_vpn_lib::command::{self, Command, Response};
+use gnosis_vpn_lib::socket;
 
-fn as_internal_cmd(cmd: &cli::Command) -> command::Command {
-    match cmd {
-        cli::Command::Status => command::Command::Status,
-        cli::Command::EntryNode {
-            endpoint,
-            api_token,
-            listen_host,
-            hop,
-            intermediate_id,
-        } => command::Command::EntryNode {
-            endpoint: endpoint.clone(),
-            api_token: api_token.clone(),
-            listen_host: listen_host.clone(),
-            hop: *hop,
-            intermediate_id: *intermediate_id,
-        },
-        cli::Command::ExitNode { peer_id } => command::Command::ExitNode { peer_id: *peer_id },
-    }
-}
+mod cli;
 
 fn main() {
-    // install global collector configured based on RUST_LOG env var.
-    tracing_subscriber::fmt::init();
+    let args = cli::parse();
 
-    let options = cli::cli().run();
+    let cmd: Command = args.command.into();
+    let resp = match socket::process_cmd(&args.socket_path, &cmd) {
+        Ok(resp) => resp,
+        Err(e) => {
+            eprintln!("Error processing {}: {}", cmd, e);
+            return;
+        }
+    };
 
-    tracing::debug!(?options, "Options parsed");
+    if args.json {
+        json_print(&resp)
+    } else {
+        pretty_print(&resp)
+    };
 
-    for cmd in options.commands.into_iter() {
-        let cmd = as_internal_cmd(&cmd);
-        let res = socket::process_cmd(&cmd);
-        match res {
-            Ok(socket::ReturnValue::WithResponse(s)) => tracing::info!("{} responded with: {}", cmd, s),
-            Ok(_) => tracing::info!("{} executed successfully", cmd),
-            Err(x) => tracing::warn!("{} failed with: {:?}", cmd, x),
+    let exit = determine_exitcode(&resp);
+    process::exit(exit);
+}
+
+fn json_print(resp: &Response) {
+    match serde_json::to_string_pretty(resp) {
+        Ok(s) => println!("{}", s),
+        Err(e) => eprintln!("Error serializing response to JSON: {}", e),
+    }
+}
+
+fn pretty_print(resp: &Response) {
```

Comment:
You should imeoplement `Debug` and `Display` on the `Response` object and then just print those.

Reply:
Yeah I could, but in this case I think it is not as straightforward.
Some of the responses should go to stderr others to sdtout. Also the printout is specifically tailored towards being good readable and formatted on the CLI. Not something I would expect on the Display trait.